cranelift/codegen/meta/src/shared/instructions.rs

```diff
@@ -4592,14 +4591,124 @@ pub(crate) fn define(
         .other_side_effects(true),
     );
 
+    ig.push(
+        Inst::new(
+            "atomic_uload8",
+            r#"
+        Atomically load 8 bits from memory at `p` and zero-extend to either 32 or 64 bits.
+
+        This is equivalent to ``load.i8`` followed by ``uextend``.
```

Comment:
Is there any reason we can't instead pattern-match a `uextend` on an `atomic_load` result?

The general rule we've tried to stick to (see some recent discussion [here](https://github.com/bytecodealliance/wasmtime/pull/2982#discussion_r657539856)) is that if we have a behavior that is just the combination of two opcodes, we should pattern-match on the primitives rather than introduce a new combo op at the CLIF level. There are a number of advantages to this approach in terms of simplicity, generality, approachability, and maintainability; the main downside is that it's a little more work at lowering time.

Reply:
I feel if we have instructions that virtually every backend will perform the same pattern matching and combinations on, then it makes sense to have it in the IR. It originally seems to have been a big enough downside that no backend was pattern matching and instead just made assumptions about how atomic load was used, which isn't great! But if not separate instructions, is there a way that we can attach some form of extension/truncation attribute on the load and store instructions? This could be a nice way of refactoring all the memory operations.